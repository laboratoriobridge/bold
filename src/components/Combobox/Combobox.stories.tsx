import { action } from '@storybook/addon-actions'
import { boolean, number, optionsKnob, select, text } from '@storybook/addon-knobs'
import matchSorter from 'match-sorter'
import React from 'react'
import { useTheme } from '../../styles'
import { Button } from '../Button'
import { HFlow } from '../HFlow'
import { Text } from '../Text'
import { Combobox } from './Combobox'
import { ComboboxMenuItem } from './ComboboxMenuComponents'

type Fruit = { value: number; label: string }

const fruits: Fruit[] = [
  { value: 1, label: 'Apple' },
  { value: 2, label: 'Avocado' },
  { value: 3, label: 'Banana' },
  { value: 4, label: 'Blueberry' },
  { value: 5, label: 'Coconut' },
  { value: 6, label: 'Grape' },
  { value: 7, label: 'Lemon' },
  { value: 8, label: 'Mango' },
  { value: 9, label: 'Melon' },
  { value: 10, label: 'Orange' },
  { value: 11, label: 'Peach' },
  { value: 12, label: 'Pear' },
]

const loadFruitsAsync = (query: string): Promise<Fruit[]> => {
  action('items loaded')()
  return new Promise((resolve) => {
    setTimeout(
      () =>
        resolve(
          matchSorter<Fruit>(fruits, query, { keys: [(item) => item.label] })
        ),
      1000
    )
  })
}

function CustomComponent(props: React.HTMLAttributes<HTMLDivElement>) {
  const theme = useTheme()

  return (
    <div
      style={{
        background: theme.pallete.surface.background,
        padding: '0.25rem 0.5rem',
        cursor: 'initial',
      }}
      {...props}
    />
  )
}

export default {
  title: 'Components/Combobox',
}

export const Default = () => (
  <Combobox<Fruit>
    value={fruits.find((e) => e.label === select('value', ['', ...fruits.map((e) => e.label)], ''))}
    label='Fruit'
    name='fruit'
    items={fruits}
    error={text('error', '')}
    menuMinWidth={number('menuMinWidth (px)', undefined)}
    itemToString={(item) => item?.label}
    placeholder={text('placeholder', 'Select a value...')}
    clearable={boolean('clearable', true)}
    disabled={boolean('disabled', false)}
    openOnFocus={boolean('openOnFocus', true)}
    loading={boolean('loading', false)}
    onChange={action('changed')}
    onFilterChange={action('filter changed')}
    onBlur={action('blur')}
    onFocus={action('focus')}
  />
)

export const Suggestion = () => (
  <Combobox<Fruit>
    createNewItem={(str) => ({ value: Math.random(), label: str })}
    label='Fruit'
    name='fruit'
    items={fruits}
    error={text('error', '')}
    icon={null}
    itemToString={(item) => item && item.label}
    placeholder='Select a value...'
    clearable={boolean('clearable', true)}
    disabled={boolean('disabled', false)}
    openOnFocus={boolean('openOnFocus', false)}
    loading={boolean('loading', false)}
    onChange={action('changed')}
    onBlur={action('blur')}
    onFocus={action('focus')}
  />
)

export const Async = () => (
  <Combobox<Fruit>
    value={fruits.find((e) => e.label === select('value', ['', ...fruits.map((e) => e.label)], ''))}
    label='Fruit Repository'
    name='repository'
    items={loadFruitsAsync}
    error={text('error', '')}
    itemToString={(item) => item && item.label}
    placeholder='Select a value...'
    clearable={boolean('clearable', true)}
    disabled={boolean('disabled', false)}
    openOnFocus={boolean('openOnFocus', true)}
    onChange={action('changed')}
    onBlur={action('blur')}
    onFocus={action('focus')}
  />
)

export const CustomComponents = () => (
  <Combobox<Fruit>
    label='Fruit'
    name='fruit'
    items={fruits}
    createNewItem={boolean('createNewItem', false) && (() => fruits[0])}
    itemToString={(item) => item && item.label}
    components={{
      Item: (props) => (
        <ComboboxMenuItem {...props}>
          <Text color='success'>Custom {props.itemToString(props.item)}</Text>
        </ComboboxMenuItem>
      ),
      PrependItem: (props) => <CustomComponent>Prepend item</CustomComponent>,
      EmptyItem: (props) => <CustomComponent>Empty item</CustomComponent>,
      CreateItem: (props) => <CustomComponent>Create item</CustomComponent>,
      LoadingItem: (props) => <CustomComponent>Loading item...</CustomComponent>,
      AppendItem: (props) => (
        <CustomComponent>
          <HFlow alignItems='center' justifyContent='space-between'>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipisicing elit. Maxime quod modi, inventore quasi aut sed beatae
              corrupti repellendus minima voluptatem debitis, quibusdam repudiandae totam voluptatum odit.
            </Text>
            <Button kind='primary' size='small' onClick={action('New item click')}>
              New item
            </Button>
          </HFlow>
        </CustomComponent>
      ),
    }}
  />
)

export const MultiSelect = () => {
  const selectedValues = optionsKnob('values', Object.fromEntries(fruits.map((e) => [e.label, e.label])), [], {
    display: 'multi-select',
  })

  return (
    <Combobox<Fruit>
      itemIsEqual={(a, b) => a.label === b.label}
      value={fruits.filter((e) => selectedValues.includes(e.label))}
      label='Fruit'
      name='fruit'
      items={fruits}
      error={text('error', '')}
      menuMinWidth={number('menuMinWidth (px)', undefined)}
      itemToString={(item) => item?.label}
      placeholder={text('placeholder', 'Select the values...')}
      clearable={boolean('clearable', true)}
      disabled={boolean('disabled', false)}
      openOnFocus={boolean('openOnFocus', true)}
      onChange={action('changed')}
      onFilterChange={action('filter changed')}
      onBlur={action('blur')}
      onFocus={action('focus')}
      multiple
    />
  )
}

export const Inline = () => (
  <Combobox<Fruit>
    inline
    defaultButtonText={text('defaultButtonText', 'Fruit')}
    value={fruits.find((e) => e.label === select('value', ['', ...fruits.map((e) => e.label)], ''))}
    name='fruit'
    items={fruits}
    error={text('error', '')}
    menuMinWidth={number('menuMinWidth (px)', undefined)}
    itemToString={(item) => item?.label}
    searchInputPlaceholder={text('searchInputPlaceholder', 'Search...')}
    disabled={boolean('disabled', false)}
    showSearchInput={boolean('showSearchInput', true)}
    loading={boolean('loading', false)}
    onChange={action('changed')}
    onFilterChange={action('filter changed')}
    onBlur={action('blur')}
    onFocus={action('focus')}
  />
)
