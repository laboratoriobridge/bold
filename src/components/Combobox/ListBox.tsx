import React, { CSSProperties, forwardRef, HTMLAttributes, MouseEvent, Ref } from 'react'
import { UseSelectPropGetters } from 'downshift'
import { Theme, useStyles } from '../../styles'
import { ComboboxComponents } from './ComboboxMenuComponents'

interface ListBoxProps<T> extends HTMLAttributes<HTMLDivElement> {
  components: ComboboxComponents<T>
  items: T[]
  loading: boolean
  highlightedIndex: number
  itemToString(item: T): string
  createNewItem?(inputValue: string): T
  getItemProps(options: { index: number; item: T }): ReturnType<UseSelectPropGetters<T>['getItemProps']>
  isItemSelected?(item: T): boolean
  onItemClick?(item: T, ev: MouseEvent<HTMLLIElement>): void
}

export const ListBox = forwardRef(ListBoxInner) as <T>(
  props: ListBoxProps<T> & { ref?: Ref<HTMLDivElement> }
) => ReturnType<typeof ListBoxInner>

function ListBoxInner<T>(props: ListBoxProps<T>, ref: Ref<HTMLDivElement>) {
  const {
    items,
    components,
    loading,
    highlightedIndex,
    getItemProps,
    itemToString,
    createNewItem,
    isItemSelected,
    onItemClick,
    tabIndex,
    ...rest
  } = props
  const { classes } = useStyles(createStyles)

  const { CreateItem, AppendItem, EmptyItem, Item, LoadingItem, PrependItem } = components

  return (
    <div ref={ref} tabIndex={tabIndex} {...rest}>
      <ul className={classes.list} tabIndex={tabIndex}>
        {PrependItem && <PrependItem />}
        {loading && <LoadingItem />}
        {!loading && createNewItem && !items?.length && <CreateItem />}
        {!loading && !createNewItem && !items?.length && <EmptyItem />}
        {items.map((item, index) => {
          const { onClick, ...itemProps } = getItemProps({ item, index })

          return (
            <Item
              key={`${item}${index}`}
              item={item}
              index={index}
              highlighted={highlightedIndex === index}
              itemToString={itemToString}
              selected={!!isItemSelected?.(item)}
              onClick={(...args) => {
                onItemClick ? onItemClick?.(item, ...args) : onClick?.(...args)
              }}
              {...itemProps}
            />
          )
        })}
        {AppendItem && <AppendItem />}
      </ul>
    </div>
  )
}

export const createStyles = (theme: Theme) => ({
  list: {
    border: 0,
    borderRadius: 0,
    boxShadow: 'none',
    maxHeight: 'auto',
    listStyle: 'none',
    margin: 0,
    padding: 0,
    backgroundColor: theme.pallete.surface.main,
    overflowY: 'auto',
    overflowX: 'hidden',
    width: '100%',
  } as CSSProperties,
})
