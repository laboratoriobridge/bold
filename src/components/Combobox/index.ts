export { Combobox, ComboboxProps } from './Combobox'
