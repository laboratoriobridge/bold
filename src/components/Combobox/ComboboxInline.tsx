import { useSelect } from 'downshift'
import matchSorter from 'match-sorter'
import React, { ChangeEvent, CSSProperties, useCallback, useEffect, useRef, useState } from 'react'
import { PopperProps, usePopper } from 'react-popper'
import { useMemo } from 'react'
import { isNil } from 'lodash'
import { Theme, useStyles } from '../../styles'
import { composeHandlers, composeRefs } from '../../util/react'
import { FormControl } from '../FormControl'
import { useFormControl, UseFormControlProps } from '../../hooks/useFormControl'
import { Button, ButtonProps } from '../Button'
import { Icon } from '../Icon'
import { Text } from '../Text'
import { ComboboxComponents, ComboboxMenuItemProps, defaultComboboxComponents } from './ComboboxMenuComponents'
import { useComboboxItemsLoader } from './useComboboxItemsLoader'
import { SearchBox } from './SearchBox'
import { ListBox } from './ListBox'

export interface ComboboxInlineProps<T>
  extends Omit<ButtonProps, 'value' | 'onChange' | 'placeholder'>,
    Omit<UseFormControlProps, 'label'> {
  value?: T
  items: T[] | ((query: string) => Promise<T[]>)
  itemToString(item: T): string
  loading: boolean
  open?: boolean
  debounceMilliseconds: number
  menuMinWidth?: number
  filter?(items: T[], filter: string): T[]
  onChange?: (newValue: T) => void
  onFilterChange?: (newValue: string) => void
  components?: Omit<Partial<ComboboxComponents<T>>, 'CreateItem'>
  popperProps?: Omit<Partial<PopperProps<any>>, 'children'>

  menuId?: string
  getItemId?(index: number): string

  defaultButtonText: string
  searchBoxPlaceholder?: string
  showSearchBox?: boolean
}

export function ComboboxInline<T>(props: ComboboxInlineProps<T>) {
  const defaultFilter = useCallback((items, filter) => matchSorter(items, filter, { keys: [props.itemToString] }), [
    props.itemToString,
  ])

  const {
    value,
    defaultButtonText,
    items,
    loading: externalLoading,
    debounceMilliseconds,
    components,
    itemToString,
    menuMinWidth = '12rem',
    onChange,
    onFocus,
    onClick,
    onBlur,
    onFilterChange,
    filter = defaultFilter,
    menuId,
    getItemId,
    error,
    searchBoxPlaceholder,
    showSearchBox = true,
    open,
    popperProps,
    ...rest
  } = props

  const [itemsLoaded, setItemsLoaded] = useState(false)
  const { classes } = useStyles(createStyles)

  const isAsync = typeof items === 'function'
  const getItems = useCallback((query: string) => (typeof items === 'function' ? items(query) : filter(items, query)), [
    items,
    filter,
  ])
  const { loading: loadingItems, items: loadedItems, loadItems } = useComboboxItemsLoader(
    getItems,
    debounceMilliseconds
  )
  const isLoading = externalLoading || (isAsync && loadingItems)

  // Reload items when changed
  useEffect(() => setItemsLoaded(false), [items])

  const toggleButtonRef = useRef<HTMLButtonElement>()
  const [searchBoxRef, setSearchBoxRef] = useState<HTMLInputElement>()
  const menuRef = useRef<HTMLDivElement>()

  useEffect(() => {
    if (open && !itemsLoaded) {
      loadItems(searchBoxRef?.value)
      setItemsLoaded(true)
    }
  }, [open, itemsLoaded, loadItems, searchBoxRef])

  const {
    selectedItem,
    isOpen,
    highlightedIndex,
    getLabelProps,
    getMenuProps,
    getToggleButtonProps,
    getItemProps,
    closeMenu,
  } = useSelect<T>({
    selectedItem: value,
    items: loadedItems,

    itemToString,
    onSelectedItemChange: ({ selectedItem }) => {
      closeMenu()
      onChange?.(selectedItem)
    },
    onIsOpenChange: ({ isOpen }) => {
      isOpen && !itemsLoaded && loadItems(null)
      setItemsLoaded(isOpen)
    },

    menuId,
    getItemId,

    ...(isNil(open) ? {} : { isOpen: open }),
  })
  const onSearchBoxValueChange = useCallback(
    ({ target: { value: inputValue } }: ChangeEvent<HTMLInputElement>) =>
      composeHandlers(loadItems, onFilterChange)(inputValue),
    [loadItems, onFilterChange]
  )

  const { getFormControlProps } = useFormControl(props)
  const formControlProps = getFormControlProps()
  const invalid = !!formControlProps.error
  const { id: labelId, htmlFor, ...downshiftLabelProps } = getLabelProps()
  const { onBlur: menuOnBlur, ...downshiftMenuProps } = getMenuProps({
    tabIndex: -1,
    'aria-invalid': invalid,
    'aria-errormessage': formControlProps.errorId,
  })
  const { ref: downshiftToggleButtonRef, ...downshiftToggleButtonProps } = getToggleButtonProps({
    onClick,
    onFocus,
    onBlur,
    'aria-controls': downshiftMenuProps.id,
    'aria-labelledby': labelId,
  })

  const {
    styles: { popper: popperStyles },
    attributes: { popper: popperAttributes },
  } = usePopper(toggleButtonRef.current, menuRef.current, {
    placement: 'bottom-start',
    ...popperProps,
  })

  useEffect(() => {
    isOpen && searchBoxRef?.focus()
  }, [isOpen, searchBoxRef])

  const componentsInner = useMemo(
    () => ({
      ...defaultComboboxComponents,
      ...(components ?? {}),
      PrependItem: (props: ComboboxMenuItemProps) => (
        <>
          {showSearchBox && (
            <div className={classes.searchBoxContainer}>
              <SearchBox ref={setSearchBoxRef} placeholder={searchBoxPlaceholder} onChange={onSearchBoxValueChange} />
            </div>
          )}
          {components?.PrependItem && <components.PrependItem {...props} />}
        </>
      ),
    }),
    [components, showSearchBox, searchBoxPlaceholder, onSearchBoxValueChange, classes.searchBoxContainer]
  )

  return (
    <div>
      <FormControl {...formControlProps}>
        <Button
          innerRef={composeRefs(toggleButtonRef, downshiftToggleButtonRef)}
          skin='ghost'
          kind={invalid ? 'danger' : 'normal'}
          size='small'
          {...downshiftToggleButtonProps}
          {...rest}
        >
          {/*By the ARIA definition, the label element is mandatory*/}
          <Text
            component='label'
            style={selectedItem != null ? { display: 'none' } : { cursor: 'pointer' }}
            id={labelId}
            {...downshiftLabelProps}
            color={invalid ? 'danger' : 'normal'}
          >
            {defaultButtonText}
          </Text>
          {selectedItem != null && <Text color={invalid ? 'danger' : 'normal'}>{itemToString(selectedItem)}</Text>}
          <Icon style={{ marginLeft: '0.5rem' }} icon={isOpen ? 'angleUp' : 'angleDown'} />
        </Button>
      </FormControl>

      {/*By the ARIA definition, the menu element should always be in the DOM*/}
      <div aria-busy={isLoading} {...downshiftMenuProps}>
        {isOpen && (
          <ListBox<T>
            ref={menuRef}
            data-testid='menu'
            className={classes.menu}
            style={{
              ...popperStyles,
              width: toggleButtonRef.current?.clientWidth,
              minWidth: menuMinWidth,
            }}
            {...popperAttributes}
            components={componentsInner}
            getItemProps={getItemProps}
            highlightedIndex={highlightedIndex}
            itemToString={itemToString}
            items={loadedItems}
            loading={isLoading}
          />
        )}
      </div>
    </div>
  )
}

export const createStyles = (theme: Theme) => ({
  searchBoxContainer: {
    padding: '0.5rem',
    borderBottom: `1px solid ${theme.pallete.divider}`,
  },

  menu: {
    display: 'flex',
    flexDirection: 'column',
    zIndex: theme.zIndex.dropdown,
    border: `1px solid ${theme.pallete.divider}`,
    borderRadius: theme.radius.popper,
    backgroundColor: theme.pallete.surface.main,
    boxShadow: theme.shadows.outer['40'],
    maxHeight: '20rem',
  } as CSSProperties,
})
