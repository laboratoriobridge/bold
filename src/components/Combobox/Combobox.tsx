import React from 'react'
import { ComboboxInline, ComboboxInlineProps } from './ComboboxInline'
import { ComboboxMultiselect, ComboboxMultiselectProps } from './ComboboxMultiselect'
import { ComboboxSingleselect, ComboboxSingleselectProps } from './ComboboxSingleselect'

export interface DefaultComboboxItemType {
  value: any
  label: string
}

export type ComboboxProps<T> =
  | (ComboboxInlineProps<T> & { readonly multiple: false; readonly inline: true })
  | (ComboboxMultiselectProps<T> & { readonly multiple: true; readonly inline: false })
  | (ComboboxSingleselectProps<T> & { readonly multiple: false; readonly inline: false })

export function Combobox<T = DefaultComboboxItemType>({ inline, multiple, ...props }: ComboboxProps<T>) {
  if (inline === true) return <ComboboxInline {...(props as ComboboxInlineProps<T>)} />
  else if (multiple === true) return <ComboboxMultiselect {...(props as ComboboxMultiselectProps<T>)} />
  else return <ComboboxSingleselect {...(props as ComboboxSingleselectProps<T>)} />
}

Combobox.defaultProps = {
  loading: false,
  debounceMilliseconds: 350,
  multiple: false,
  inline: false,
} as Partial<ComboboxProps<any>>
