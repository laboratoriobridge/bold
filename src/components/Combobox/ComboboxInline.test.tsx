import { act, fireEvent, render } from '@testing-library/react'
import { resetIdCounter } from 'downshift'
import React from 'react'

import { DefaultItemType } from '../Select/SelectSingle'

import { ComboboxInline, ComboboxInlineProps } from '../Combobox/ComboboxInline'

const items: DefaultItemType[] = [
  { label: 'Value #001', value: 1 },
  { label: 'Value #002', value: 2 },
  { label: 'Value #003', value: 3 },
  { label: 'Value #004', value: 4 },
]

const itemToString = (item: DefaultItemType) => item && item.label

const ComboboxInlineTest = (props?: Partial<ComboboxInlineProps<DefaultItemType>>) => (
  <ComboboxInline<DefaultItemType>
    value={items[0]}
    items={items}
    itemToString={itemToString}
    defaultButtonText='ComboboxInline'
    searchInputPlaceholder='Search for a value'
    openOnFocus
    loading={false}
    debounceMilliseconds={500}
    {...props}
  />
)

beforeEach(() => resetIdCounter())

describe('ComboboxInline', () => {
  it('should render correctly when closed', () => {
    const { container } = render(<ComboboxInlineTest />)
    expect(container).toMatchSnapshot()
  })
  it('should render correctly when opened', async () => {
    let container: HTMLElement
    await act(async () => {
      const result = render(<ComboboxInlineTest />)
      container = result.container
    })
    const button = container.querySelector('button')
    await act(async () => {
      fireEvent.click(button)
    })
    expect(document.body).toMatchSnapshot()
  })
  it('should make the popper content visible on click', async () => {
    let container: HTMLElement
    await act(async () => {
      const result = render(<ComboboxInlineTest />)
      container = result.container
    })
    const button = container.querySelector('button')
    expect(container.querySelector('ul')).toBeFalsy()
    await act(async () => {
      fireEvent.click(button)
    })
    expect(container.querySelector('ul')).toBeTruthy()
  })

  it('should focus the input field when opened', async () => {
    let container: HTMLElement
    await act(async () => {
      const result = render(<ComboboxInlineTest />)
      container = result.container
    })
    const button = container.querySelector('button')
    await act(async () => {
      fireEvent.click(button)
    })
    expect(document.activeElement).toEqual(container.querySelector('input'))
  })
  it('should contain search input box', async () => {
    let container: HTMLElement
    await act(async () => {
      const result = render(<ComboboxInlineTest />)
      container = result.container
    })
    const button = container.querySelector('button')
    await act(async () => {
      fireEvent.click(button)
    })
    expect(container.querySelector('input')).not.toBeNull()
  })
})
