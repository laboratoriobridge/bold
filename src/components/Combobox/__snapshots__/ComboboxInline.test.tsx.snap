// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboboxInline should render correctly when closed 1`] = `
.emotion-2 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5rem;
  color: #24252E;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #24252E;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
}

.emotion-2 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-2:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0069D0;
}

.emotion-2:not(:disabled):hover {
  background-color: rgba(58,58,71,0.1);
}

.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  color: #24252E;
}

.emotion-1 {
  fill: currentColor;
  font-size: 1.5rem;
  margin-left: 0.5rem;
}

<div>
  <button
    class="emotion-2"
    type="button"
  >
    <span>
      <span
        class="emotion-0"
      >
        Value #001
      </span>
      <svg
        aria-hidden="true"
        class="emotion-1"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          clip-rule="evenodd"
          d="M16.294 8.294a1 1 0 111.414 1.414l-5 5a1 1 0 01-1.414 0l-5.001-5a1 1 0 011.414-1.415l4.294 4.294 4.293-4.293z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`ComboboxInline should render correctly when opened 1`] = `
.emotion-2 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5rem;
  color: #24252E;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #24252E;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
}

.emotion-2 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-2:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0069D0;
}

.emotion-2:not(:disabled):hover {
  background-color: rgba(58,58,71,0.1);
}

.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  color: #24252E;
}

.emotion-1 {
  fill: currentColor;
  font-size: 1.5rem;
  margin-left: 0.5rem;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  z-index: 1300;
  border: 1px solid #D3D4DD;
  border-radius: 2px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 4px -1px rgba(0,0,0,0.2),0 1px 10px 0 rgba(0,0,0,0.12),0 4px 5px 0 rgba(0,0,0,0.14);
  max-height: 20rem;
}

.emotion-11 {
  z-index: auto;
  border: 0;
  border-radius: 0;
  box-shadow: none;
  max-height: auto;
  list-style: none;
  margin: 0;
  padding: 0;
  background-color: #FFFFFF;
  overflow-y: auto;
  overflow-x: hidden;
  width: 100%;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-8 {
  display: inline-block;
  margin-bottom: 0rem;
  margin-top: 0rem;
}

.emotion-8:first-of-type {
  margin-top: 0;
}

.emotion-8:last-of-type {
  margin-bottom: 0;
}

.emotion-8:empty {
  display: none;
}

.emotion-6 {
  position: relative;
}

.emotion-6 input {
  padding-left: 2.5rem;
}

.emotion-5 {
  position: absolute;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.25rem;
  width: 2.5rem;
  top: 1px;
  bottom: 1px;
  left: 1px;
}

.emotion-3 {
  background-color: #FFFFFF;
  border: solid 1px #8F8FA2;
  border-radius: 2px;
  color: #24252E;
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1rem;
  margin: 0;
  padding: calc(0.5rem - 1px) 0.5rem;
  width: 100%;
  -webkit-transition-property: box-shadow;
  transition-property: box-shadow;
  -webkit-transition-duration: .2s;
  transition-duration: .2s;
}

.emotion-3[type="search"]::-webkit-search-decoration {
  display: none;
}

.emotion-3[type="search"]::-webkit-search-cancel-button {
  display: none;
}

.emotion-3[type="search"]::-webkit-search-results-button {
  display: none;
}

.emotion-3[type="search"]::-webkit-search-results-decoration {
  display: none;
}

.emotion-3:required {
  box-shadow: none;
}

.emotion-3::-webkit-input-placeholder {
  color: #696979;
}

.emotion-3::-moz-placeholder {
  color: #696979;
}

.emotion-3:-ms-input-placeholder {
  color: #696979;
}

.emotion-3::placeholder {
  color: #696979;
}

.emotion-3:disabled {
  border-color: #D3D4DD;
  background-color: #F0F0F5;
}

.emotion-3:not(:disabled):hover {
  border-color: #7A7B8E;
}

.emotion-3:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-3:not(:disabled):active {
  border-color: #0069D0;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
}

.emotion-4 {
  fill: currentColor;
  font-size: 1.5rem;
  border-radius: inherit;
}

.emotion-4:focus {
  box-shadow: none;
}

.emotion-10 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  border-bottom: 1px solid #D3D4DD;
  cursor: pointer;
  padding: 0.5rem 0.5rem;
  -webkit-transition: .1s ease;
  transition: .1s ease;
}

.emotion-10:last-of-type {
  border-bottom: none;
}

.emotion-10:hover {
  background: #F0F0F5;
}

.emotion-10:focus {
  outline: 0;
  border-radius: 3px;
  box-shadow: inset 0 0 0 2px #0069D0;
}

.emotion-13 {
  z-index: 1200;
  border: 1px solid #D3D4DD;
  box-shadow: 0 2px 4px -1px rgba(0,0,0,0.2),0 1px 10px 0 rgba(0,0,0,0.12),0 4px 5px 0 rgba(0,0,0,0.14);
  background-color: #FFFFFF;
  padding: 0.5rem;
  padding-bottom: 0;
  position: absolute;
  top: 0;
  right: auto;
  bottom: auto;
  left: 0;
  -webkit-transform: translate(0px,0px);
  -ms-transform: translate(0px,0px);
  transform: translate(0px,0px);
}

.emotion-13 [role=listbox] {
  width: calc(100% + 1rem);
  margin-left: -0.5rem;
  margin-top: 0.5rem;
}

.emotion-13 [role=listbox] > div {
  border-width: 1px 0 0 0;
  box-shadow: none;
}

<body>
  <div>
    <button
      class="emotion-2"
      type="button"
    >
      <span>
        <span
          class="emotion-0"
        >
          Value #001
        </span>
        <svg
          aria-hidden="true"
          class="emotion-1"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            clip-rule="evenodd"
            d="M7.707 14.707a1 1 0 11-1.414-1.414l5-5a1 1 0 011.414 0l5.001 5.001a1 1 0 01-1.414 1.414L12 10.414l-4.293 4.293z"
          />
        </svg>
      </span>
    </button>
    <div
      class="emotion-13"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
    >
      <div
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="downshift-0-menu"
        role="combobox"
      >
        <div
          class="emotion-9"
        >
          <div
            class="emotion-8"
          >
            <div
              class="emotion-7"
            >
              <div
                class="emotion-6"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="downshift-0-menu"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  class="emotion-3"
                  id="downshift-0-input"
                  placeholder="Search for a value"
                  type="text"
                  value=""
                />
                <span
                  class="emotion-5"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="Hide options"
                    class="emotion-4"
                    height="1em"
                    role="img"
                    viewBox="0 0 24 24"
                    width="1em"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M14.605 13.194l7 6.998a1 1 0 01-1.414 1.414h-.001l-6.997-7a7 7 0 111.412-1.412zM9 14A5 5 0 109 4a5 5 0 000 10z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-busy="false"
          aria-labelledby="downshift-0-label"
          id="downshift-0-menu"
          role="listbox"
        >
          <div
            class="emotion-12"
            data-testid="menu"
          >
            <ul
              class="emotion-11"
            >
              <li
                class="emotion-10"
              >
                No results were found
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
